core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java

```diff
@@ -0,0 +1,106 @@
+import static org.mule.tck.probe.PollingProber.check;
+
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+
+import org.mule.runtime.api.component.location.ComponentLocation;
+import org.mule.runtime.api.util.concurrent.Latch;
+import org.mule.runtime.core.privileged.event.BaseEventContext;
+import org.mule.tck.junit4.AbstractMuleTestCase;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+
+import io.qameta.allure.Issue;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class AsyncEventContextCompletionTestCase extends AbstractMuleTestCase {
+
+  private ExecutorService executorService;
+
+  @Before
+  public void setup() {
+    executorService = newCachedThreadPool();
+  }
+
+  @After
+  public void after() {
+    if (executorService != null) {
+      executorService.shutdown();
+    }
+  }
+
+  @Test
+  @Issue("W-16640678")
+  public void childContextCompletesAsyncAfterParentContext() throws Exception {
+    final CompletableFuture<Void> parentTermination = new CompletableFuture<>();
+    final Latch testLatch = new Latch();
+    final Latch parentCompletedLatch = new Latch();
+    final Latch childCompletedLatch = new Latch();
+    final Latch childTerminatedLatch = new Latch();
+
+    BaseEventContext parentContext =
+        (BaseEventContext) create("id", "server", mock(ComponentLocation.class), "correlation", of(parentTermination));
+    BaseEventContext childContext = DefaultEventContext.child(parentContext, empty());
+
+
+    parentContext.onComplete((e, t) -> {
+      parentCompletedLatch.countDown();
+      await(childCompletedLatch);
+      await(childTerminatedLatch);
+      executorService.submit(() -> parentTermination.complete(null));
+    });
+
+    parentContext.onTerminated((e, t) -> executorService.submit(() -> {
+      testLatch.release();
+    }));
+
+    childContext.onComplete((e, t) -> {
+      executorService.submit(() -> {
+        parentContext.success();
+      });
+      await(parentCompletedLatch);
+      childCompletedLatch.release();
+    });
+
+    childContext.onTerminated((e, t) -> childTerminatedLatch.countDown());
+
+    executorService.submit(() -> childContext.success());
+
+    await(testLatch);
+
+    check(5000, 500, () -> {
+      assertThat(parentContext.isTerminated(), is(true));
+      assertThat(childContext.isTerminated(), is(true));
+      return true;
+    });
+  }
+
+  private void await(CountDownLatch latch) {
+    try {
+      if (!latch.await(1000000000, SECONDS)) {
+        throw new AssertionError("Latch timed out");
```

Comment:
the test will timeout before this: just use the no timeout version, or put a meaningful lower value (30 secs?)

Reply:
this was this long for debugging purposes and forgot to change it.